core/domain/html_cleaner.py

```diff
@@ -782,11 +783,202 @@ def add_caption_attr_to_image(html_string):
+                        soup_for_tabs, err_dict, html_string)
+            except Exception:
+                if 'oppia-noninteractive-tabs' not in err_dict:
+                    err_dict['oppia-noninteractive-tabs'] = html_string
+                elif html_string not in err_dict[
+                        'oppia-noninteractive-tabs']:
+                    err_dict['oppia-noninteractive-tabs'].append(html_string)
+                break
+
+    return err_dict
+
+
+def _validate_customization_args_in_soup(soup, err_dict, html_string):
+    """Validates customization arguments of Rich Text Components in a soup.
+
+    Args:
+        soup: bs4.BeautifulSoup. The html soup in which customization arguments
+            are to be validated.
+        err_dict: dict. Dictionary of all the invalid customisation args where
+            key is a Rich Text Component and value is the invalid html string.
+        html_string: str. The html string to be stored in dict if it is invalid.
+    """
+    # Dictionary to hold html strings in which customization arguments
+    # are invalid.
+    # Importing this at the top of the file causes a circular dependency.
+    # So this is needed here.
+    from extensions.objects.models import objects # pylint: disable=relative-import
+
+    rich_text_components = feconf.RICH_TEXT_COMPONENTS
+    # Mapping of type to classes defined in objects.py.
+    class_dict = {
+        'SanitizedUrl': objects.SanitizedUrl,
+        'UnicodeString': objects.UnicodeString,
+        'MathLatexString': objects.MathLatexString,
+        'Filepath': objects.Filepath,
+        'Int': objects.Int,
+        'Boolean': objects.Boolean,
+        'Html': objects.Html
+    }
+    filename_re = r'^[A-Za-z0-9+/]*\.((png)|(jpeg)|(gif)|(jpg))$'
+
+    for component in rich_text_components:
+        # This flag is set to true if the html string is invalid.
+        is_invalid = False
+        tag_name = component['name']
+        customization_args = component['customization_args']
+        required_attrs_names = [arg['name'] for arg in customization_args]
+
+        for tag in soup.findAll(name=tag_name):
+            attrs = tag.attrs
+            attrs_names = attrs.keys()
+            # Ensure that no tag attribute is missing and also check
+            # there is no extra attribute.
+            if set(attrs_names) != set(required_attrs_names):
+                is_invalid = True
+                break
+
+            for customization_arg in customization_args:
+                arg_name = customization_arg['name']
+                arg_value = json.loads(unescape_html(attrs[arg_name]))
+                if customization_arg['type'] != 'custom':
+                    arg_type = class_dict[customization_arg['type']]
+                    try:
+                        arg_type.normalize(arg_value)
+                        # To ensure filepath is in valid format.
+                        if arg_name == 'filepath-with-value':
```

Comment:
Hm this is interesting. So it seems like you're throwing all the different arg names into a single validation flow, which is ... hard to follow, and leaves this susceptible to different RTE components reusing the same arg names in different ways.

Could these validations be brought into the object-validation code itself? Can they be based on knowing the type alone (rather than relying on the arg name which really is somewhat arbitrary and can be changed in the future via migrations, as e.g. @ishucr7 is doing now)?

Reply:
Since these checks are very specific, we cannot bring them to object validations. For example,

- Every unicode need not have 11 characters, it is specific to `video-id`.
- Every filename need not end in the specified extension for images, it is specific to `filepath` in images.

Another option is to change the if condition to check tag name and argument type which will ensure that checks are not to be changed even though the attribute names are changed.
```
if tag_name == 'oppia-noninteractive-image' and arg_type == objects.Filepath:
      check something
```
What do you suggest?